ddtrace/contrib/openai/_endpoint_hooks.py

```diff
@@ -0,0 +1,265 @@
+
+        return
+
+    def _post_response(self, pin, integration, span, args, kwargs, resp, error):
+        if resp and not kwargs.get("stream"):
+            if "choices" in resp:
+                choices = resp["choices"]
+                span.set_tag("openai.response.choices.num", len(choices))
+                for choice in choices:
+                    idx = choice["index"]
+                    if "finish_reason" in choice:
+                        span.set_tag_str("openai.response.choices.%d.finish_reason" % idx, str(choice["finish_reason"]))
+                    if "logprobs" in choice:
+                        span.set_tag_str("openai.response.choices.%d.logprobs" % idx, "returned")
+                    if integration.is_pc_sampled_span(span):
+                        span.set_tag_str("openai.response.choices.%d.text" % idx, integration.trunc(choice.get("text")))
+            integration.record_usage(span, resp.get("usage"))
+            if integration.is_pc_sampled_log(span):
+                prompt = kwargs.get("prompt", "")
+                integration.log(
+                    span,
+                    "info" if error is None else "error",
+                    "sampled %s" % self._default_name,
+                    attrs={
+                        "prompt": prompt,
+                        "choices": resp["choices"] if resp and "choices" in resp else [],
+                    },
+                )
+        return self._handle_response(pin, span, integration, resp)
+
+
+class _ChatCompletionHook(_BaseCompletionHook):
+    _request_tag_attrs = [
+        "temperature",
+        "top_p",
+        "n",
+        "stream",
+        "stop",
+        "max_tokens",
+        "presence_penalty",
+        "frequency_penalty",
+        "logit_bias",
+        "user",
+    ]
+    _default_name = "chat.completions"
+
+    def _pre_response(self, pin, integration, span, args, kwargs):
+        sample_pc_span = integration.is_pc_sampled_span(span)
+        messages = kwargs.get("messages")
+        if sample_pc_span and messages:
+            for idx, m in enumerate(messages):
+                content = integration.trunc(m.get("content", ""))
+                role = integration.trunc(m.get("role", ""))
+                span.set_tag_str("openai.request.messages.%d.content" % idx, content)
+                span.set_tag_str("openai.request.messages.%d.role" % idx, role)
+
+        if "stream" in kwargs and kwargs["stream"]:
+            # streamed responses do not have a usage field, so we have to
+            # estimate the number of tokens returned.
+            est_num_message_tokens = 0
+            for m in messages:
+                est_num_message_tokens += _est_tokens(m.get("content", ""))
+            span.set_metric("openai.response.usage.prompt_tokens", est_num_message_tokens)
+
+        return
+
+    def _post_response(self, pin, integration, span, args, kwargs, resp, error):
```

Comment:
Making these functions a bit flatter (less nested) would make them easier to read and maintain.